Userland/Shell/Parser.cpp

```diff
@@ -1622,91 +1632,101 @@ RefPtr<AST::Node> Parser::parse_history_designator()
+            word_selector_kind = AST::HistorySelector::WordSelectorKind::Index;
+            offset = value.value();
+        } else if (c == '^') {
             consume();
-            return AST::HistorySelector::WordSelector {
-                AST::HistorySelector::WordSelectorKind::Index,
-                0,
-                { m_rule_start_offsets.last(), m_offset, m_rule_start_lines.last(), line() },
-                syntax_error
-            };
-        }
-        if (c == '$') {
+            word_selector_kind = AST::HistorySelector::WordSelectorKind::Index;
+            offset = 1;
+        } else if (c == '$') {
             consume();
-            return AST::HistorySelector::WordSelector {
-                AST::HistorySelector::WordSelectorKind::Last,
-                0,
-                { m_rule_start_offsets.last(), m_offset, m_rule_start_lines.last(), line() },
-                syntax_error
-            };
+            word_selector_kind = AST::HistorySelector::WordSelectorKind::Last;
+            offset = 0;
         }
-        return {};
+        if (offset == -1)
+            return {};
+        return AST::HistorySelector::WordSelector {
+            word_selector_kind,
+            static_cast<size_t>(offset),
+            { m_rule_start_offsets.last(), m_offset, m_rule_start_lines.last(), line() },
+            syntax_error
+        };
     };
 
-    auto start = parse_word_selector();
-    if (!start.has_value()) {
+    auto set_word_selector = [&](auto& word_selector, AST::HistorySelector::WordSelectorKind word_selector_kind, size_t offset) {
+        word_selector = AST::HistorySelector::WordSelector {
+            word_selector_kind,
+            offset,
+            { m_rule_start_offsets.last(), m_offset, m_rule_start_lines.last(), line() },
+            syntax_error
+        };
+    };
+
+    auto first_char = peek();
+    if (!(is_digit(first_char) || "^$-*"sv.contains(first_char))) {
         if (!syntax_error)
             syntax_error = create<AST::SyntaxError>("Expected a word selector after ':' in a history event designator", true);
-        auto node = create<AST::HistoryEvent>(move(selector));
-        node->set_is_syntax_error(*syntax_error);
-        return node;
-    }
-    selector.word_selector_range.start = start.release_value();
-
-    if (peek() == '-') {
+    } else if (first_char == '*') {
         consume();
-        auto end = parse_word_selector();
-        if (!end.has_value()) {
-            if (!syntax_error)
-                syntax_error = create<AST::SyntaxError>("Expected a word selector after '-' in a history event designator word selector", true);
-            auto node = create<AST::HistoryEvent>(move(selector));
-            node->set_is_syntax_error(*syntax_error);
-            return node;
-        }
-        selector.word_selector_range.end = move(end);
+        set_word_selector(selector.word_selector_range.start, AST::HistorySelector::WordSelectorKind::Index, 1);
```

Comment:
Let's switch this out to something like
`selector.word_selector_range.start = make_selector(...)`, it doesn't really sound like it would modify the first arg if you look at it in isolation

Reply:
Right. I will fix that.